### PR TITLE
Fix `KeyError` when having multiple routes for a single path

### PR DESCRIPTION
**Pull request by damonhook**
## Initial Issue

- Pyramid allows you to have multiple routes applied to the same path (using `predicates` or `request_method`).
    - e.g: 
    ```python
    config.add_route("todos", "/")
    ```
    vs
    ```python
    config.add_route("get_todos", "/", request_method="GET")
    config.add_route("create_todo", "/", request_method="POST")
    ```
  - Currently this will create the app without error, but cause a `KeyError` when attempting to actually call the endpoint
  - This is because the `routes` object is currently built up using a dictionary of `path` -> `route_name` (so the second route would overwrite the first).

## Changes

- Change the building of the `routes` settings object to loop through each route (by route_name) to fix the aforementioned issue (resulting `routes` object remains the same)
- Add some extra checks in the `openapi_view` deriver so that openapi can produce the notfound, rather than getting a `KeyError`, should the given route_name not be present in the `routes` object
    - Basically just check that the `route_name` exists inside the `routes` object before trying to use index it.
- Minor improvement to `remove_prefixes` in that it will actually check if it is a prefix (instead of a naïve string replace anywhere in the string)

## Question

- The building of the `routes` settings object (for multi-apis) is done at the end of the `check_all_routes` function. This means that if you disable endpoint validation (`pyramid_openapi3.enable_endpoint_validation = False`) you will not have this settings object.
    - I think we should either: Move this to its own function (also subscribing to the application created), or move it above the check against the `pyramid_openapi3.enable_endpoint_validation` setting.
    - This could be done in a separate branch / PR

### CODE REVIEW

**Comment by zupo** [comment]
Hey @damonhook, sorry for radio silence, the PR is not simple, and I haven't yet had the time to have a proper look. Hopefully by the end of this week.

File: pyramid_openapi3/__init__.py [hunk 1]
```diff
@@ -330,35 +333,47 @@ def check_all_routes(event: ApplicationCreated):
...
-                prefixes.append(path)
+        prefixes = _get_server_prefixes(openapi_settings["spec"])
 
         def remove_prefixes(path):
             path = f"/{path}" if not path.startswith("/") else path
             for prefix in prefixes:
-                path = path.replace(prefix, "")
+                if path.startswith(prefix):
+                    prefix_length = len(prefix)
+                    return path[prefix_length:]
```
Comment (zupo) [inline pyramid_openapi3/__init__.py:343]: Yeah, safer, nice!

File: pyramid_openapi3/__init__.py [hunk 2]
```diff
@@ -147,7 +149,8 @@ def explorer_view(request: Request) -> Response:
             settings = config.registry.settings
             if settings.get(apiname) is None:
                 raise ConfigurationError(
-                    "You need to call config.pyramid_openapi3_spec for explorer to work."
+                    "You need to call config.pyramid_openapi3_spec for the explorer "
+                    "to work."
```
Comment (zupo) [inline pyramid_openapi3/__init__.py:153]: I'm guessing this change is to cut the line length from 86 chars to below 80. 

This project allows 10% leeway over the 80 chars limit. The rationale is that usually one does not get a speeding ticket for going 2 miles over the limit. More in this talk: https://www.youtube.com/watch?v=wf-BqAjZb8M

File: pyramid_openapi3/__init__.py [hunk 3]
```diff
@@ -330,35 +333,47 @@ def check_all_routes(event: ApplicationCreated):
...
+        # (due to request_method predicates). So loop through each route
+        # to create a lookup of route_name -> api_name
+        for route_name, route in app.routes_mapper.routes.items():
+            if remove_prefixes(route.path) in paths:
+                settings["pyramid_openapi3"]["routes"][route_name] = name
+
+
+def _get_server_prefixes(spec: Spec) -> t.List[str]:
+    """
+    Build a set of possible route prefixes from the api spec.
```
Comment (zupo) [inline pyramid_openapi3/__init__.py:368]:
```suggestion
    """Build a set of possible route prefixes from the api spec.
```

Strange that linters didn't complain about this one. Usually we start the docstring without a new line.

**Review by zupo — APPROVED** [review]
Looks good. Can be merged.

**Comment by zupo** [comment]
I don't understand why CI is failing. Nothing that seems relevant changed in this PR and master is passing: https://app.circleci.com/pipelines/github/Pylons/pyramid_openapi3/373/workflows/21e87531-db55-4339-9528-33d8d01a630c/jobs/672.

File: pyramid_openapi3/__init__.py (hunk 2 above)
Comment (damonhook) [reply pyramid_openapi3/__init__.py:153]: Ahh, yeah it turns out my linter was not reading from the .flake8 file properly (and I have my line-length set to a strict 88), and this line was 89 characters long.

Although 89 is larger than 10% of 80 so I am unsure why `B950` wasn't complaining about it.

Comment (zupo) [reply pyramid_openapi3/__init__.py:153]: Ah no, it's over 88 chars, sorry for the noise.

File: pyramid_openapi3/__init__.py (hunk 3 above)
Comment (damonhook) [reply pyramid_openapi3/__init__.py:368]: I seem to only get the error if it is a single line docstring (D200). I am not getting the H404 error, strange.

**Comment by zupo** [comment]
Merged in https://github.com/Pylons/pyramid_openapi3/pull/120.
